pkg/controller/baremetalhost/baremetalhost_controller.go

```diff
@@ -132,6 +145,51 @@ func (r *ReconcileBareMetalHost) Reconcile(request reconcile.Request) (reconcile
 		return reconcile.Result{Requeue: true}, nil
 	}
 
+	// If we do not have all of the needed BMC credentials, set our
+	// operational status to indicate missing information.
+	if host.Spec.BMC.Address == "" {
+		reqLogger.Info(bmc.MissingAddressMsg)
+		err := r.setErrorCondition(request, host, bmc.MissingAddressMsg)
+		// Without the BMC address there's no more we can do, so we're
+		// going to return the emtpy Result anyway, and don't need to
+		// check err.
+		return reconcile.Result{}, errors.Wrap(err, "failed to set error condition")
+	}
+
+	// Load the secret containing the credentials for talking to the
+	// BMC.
+	if host.Spec.BMC.CredentialsName == "" {
+		// We have no name to use to load the secrets.
+		reqLogger.Info(bmc.MissingCredentialsMsg)
+		err := r.setErrorCondition(request, host, bmc.MissingCredentialsMsg)
+		return reconcile.Result{}, errors.Wrap(err, "failed to set error condition")
+	}
+	secretKey := host.CredentialsKey()
+	bmcCredsSecret := &corev1.Secret{}
+	err = r.client.Get(context.TODO(), secretKey, bmcCredsSecret)
+	if err != nil {
+		return reconcile.Result{}, errors.Wrap(err,
+			"failed to fetch BMC credentials from secret reference")
+	}
+	bmcCreds := bmc.Credentials{
+		Username: string(bmcCredsSecret.Data["username"]),
+		Password: string(bmcCredsSecret.Data["password"]),
+	}
+
+	// Verify that the secret contains the expected info.
+	validCreds, reason := bmcCreds.AreValid()
+	if !validCreds {
+		reqLogger.Info("invalid BMC Credentials", "reason", reason)
+		err := r.setErrorCondition(request, host, reason)
+		return reconcile.Result{}, errors.Wrap(err, "failed to set error condition")
```

Comment:
I don't understand this part.. is the error message correct?  Do we not want to return an error message about invalid creds?

Reply:
We talked about this via chat, but for other reviewers: This section stores the error state on the host object and then only throws an error if it cannot complete that save operation. The reason for that pattern is that a thrown error causes the reconciler to try again, and "bad credentials" isn't the type of error where trying again is going to do any good.